src/System.Private.CoreLib/shared/System/IO/FileStream.cs

```diff
@@ -372,6 +381,17 @@ public override ValueTask<int> ReadAsync(Memory<byte> buffer, CancellationToken
                 return new ValueTask<int>(Task.FromCanceled<int>(cancellationToken));
             }
 
+            if (!_useAsyncIO)
+            {
+                // If we weren't opened for asynchronous I/O, we still call to the base implementation so that
+                // Read is invoked asynchronously.  But if we have a byte[], we can do so using the base Stream's
+                // internal helper that bypasses delegating to BeginRead, since we already know this is FileStream
+                // rather than something derived from it and what our BeginRead implementation is going to do.
+                return MemoryMarshal.TryGetArray(buffer, out ArraySegment<byte> segment) ?
+                    new ValueTask<int>((Task<int>)base.BeginReadInternal(segment.Array!, segment.Offset, segment.Count, null, null, serializeAsynchronously: true, apm: false)) :
+                    base.ReadAsync(buffer, cancellationToken);
```

Comment:
Wouldn't we also want to use BeginReadInternal in the non-array case? The base implementation https://github.com/dotnet/coreclr/blob/3834f562e60a22e9f7feb4f12e05cf5e841257e6/src/System.Private.CoreLib/shared/System/IO/Stream.cs#L369 uses an array from ArrayPool, might need to duplicate the same logic here?

Reply:
> Wouldn't we also want to use BeginReadInternal in the non-array case?

It's not worth all the logic that would need to be duplicated changed.

Reply:
That was also my thought when I looked into it, but arguably that overload exists for performance reasons when you need to read into a buffer that isn't a managed array, so might be a bit important to get the maximal performance here (?)

Reply:
A queued work item per operation is a much larger cost. If you care about "maximal perf", you shouldn't be using the async APIs with a stream not configured for it.